File: productivity_next/productivity_next/page/timesheet_verificati/timesheet_verificati.js
frappe.pages['timesheet-verificati'].on_page_load = function(wrapper) {
    frappe.require("assets/productivity_next/css/timesheet_verificati.css");
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Timesheet Verification',
        single_column: true
    });

    // Render HTML
    $(frappe.render_template("timesheet_verificati", {})).appendTo(page.body);

    // Helper to get filter values
    function get_filters() {
        return {
            employee: $("#employee_select").val(),
            date: $("#date_select").val()
        };
    }

    // Fetch and render summary cards
    function fetch_and_render_summary() {
        frappe.call({
            method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.get_summary_data",
            args: get_filters(),
            callback: function(r) {
                let d = r.message;
                // Calls
                $("#calls-incoming-duration").text(d.calls.incoming.duration);
                $("#calls-incoming-count").text(d.calls.incoming.count);
                $("#calls-outgoing-duration").text(d.calls.outgoing.duration);
                $("#calls-outgoing-count").text(d.calls.outgoing.count);
                $("#calls-total").text(d.calls.total);

                // Meetings
                $("#meetings-internal-duration").text(d.meetings.internal.duration);
                $("#meetings-internal-count").text(d.meetings.internal.count);
                $("#meetings-external-duration").text(d.meetings.external.duration);
                $("#meetings-external-count").text(d.meetings.external.count);
                $("#meetings-total").text(d.meetings.total);

                // System
                $("#system-app-duration").text(d.system.app.duration);
                $("#system-app-count").text(d.system.app.count);
                $("#system-web-duration").text(d.system.web.duration);
                $("#system-web-count").text(d.system.web.count);
                $("#system-total").text(d.system.total);
            }
        });
    }

    // Fetch employees for filter
    frappe.call({
        method: "frappe.client.get_list",
        args: {
            doctype: "Employee",
            fields: ["name", "employee_name"],
            limit_page_length: 1000
        },
        callback: function(r) {
            let opts = r.message.map(e => `<option value="${e.name}">${e.employee_name || e.name}</option>`).join("");
            $("#employee_select").html(opts);
            // Set default to current user if possible
            frappe.call({
                method: "frappe.client.get_value",
                args: {
                    doctype: "Employee",
                    filters: { user_id: frappe.session.user },
                    fieldname: "name"
                },
                callback: function(res) {
                    if(res.message && res.message.name) {
                        $("#employee_select").val(res.message.name);
                    }
                    // Set date to today
                    $("#date_select").val(frappe.datetime.get_today());
                    fetch_and_render_summary();
                }
            });
        }
    });

    // Fetch button handler
    $(document).on("click", "#fetch_btn", function() {
    fetch_and_render_summary();
        $("#details-section").hide();
    });

    // Remove dynamic button insertion, only keep click handlers
    // Preview Timesheet handler
    $(document).on("click", "#preview_timesheet_btn", function() {
        $("#timesheet-preview-section").remove();
        frappe.call({
            method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.preview_timesheet",
            args: get_filters(),
            callback: function(r) {
                let logs = r.message.logs || [];
                let total_hours = r.message.total_hours || 0;
                let project_summary = r.message.project_summary || [];
				console.log("meg",r.message)
                let docstatus = r.message.docstatus;
                let timesheet_name = r.message.timesheet_name;
                let action_btn = "";
                if (timesheet_name) {
                    if (docstatus == 0) {
                        action_btn = `<button class='btn btn-success btn-sm me-2' id='submit_timesheet_btn' data-name='${timesheet_name}'>Submit Timesheet</button>`;
                    } else if (docstatus == 1) {
                        action_btn = `<button class='btn btn-danger btn-sm me-2' id='cancel_timesheet_btn' data-name='${timesheet_name}'>Cancel Timesheet</button>`;
                    }
                }
                let html = `<div id=\"timesheet-preview-section\" class=\"card p-4 mt-4\">
                    <div class=\"d-flex justify-content-between align-items-center mb-2\">
                        <h4 class=\"mb-0\">Timesheet Preview</h4>
                        <div>${action_btn}</div>
                    </div>
                    <div class=\"table-responsive\">
                        <table class=\"table table-bordered table-hover\">
                            <thead>
                                <tr>
                                    <th>Activity Type</th>
                                    <th>Task</th>
                                    <th>Project</th>
                                    <th>Issue</th>
                                    <th>Hours</th>
                                    <th>From Time</th>
                                    <th>To Time</th>
                                    <th>Description</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${logs.map(row => `
                                    <tr>
                                        <td>${row.activity_type || ""}</td>
                                        <td>${row.task || ""}</td>
                                        <td>${row.project || ""}</td>
                                        <td>${row.issue || ""}</td>
                                        <td>${row.hours || ""}</td>
                                        <td>${row.from_time || ""}</td>
                                        <td>${row.to_time || ""}</td>
                                        <td>${row.description || ""}</td>
                                    </tr>
                                `).join("")}
                            </tbody>
                        </table>
                    </div>
                    <div class=\"mt-3\"><strong>Total Time:</strong> ${total_hours} hours</div>
                    <div class=\"mt-3\">
                        <h5>Project-wise Time Segregation</h5>
                        <div class=\"table-responsive\">
                            <table class=\"table table-bordered table-hover\">
                                <thead>
                                    <tr>
                                        <th>Project</th>
                                        <th>Total Hours</th>
                                    </tr>
                                </thead>
                                <tbody>
                                    ${project_summary.map(row => `
                                        <tr>
                                            <td>${row.project}</td>
                                            <td>${row.total_hours}</td>
                                        </tr>
                                    `).join("")}
                                </tbody>
                            </table>
                        </div>
                    </div>
                </div>`;
                $(".timesheet-verificati-filters").after(html);
            }
        });
    });

    // Submit Timesheet handler
    $(document).on("click", "#submit_timesheet_btn", function() {
        let name = $(this).data("name");
		console.log("me",name)
        frappe.call({
            method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.submit_timesheet",
            args: {name},
            callback: function() {
                frappe.msgprint("Timesheet submitted.");
                $("#preview_timesheet_btn").click();
            }
        });
    });
    // Cancel Timesheet handler
    $(document).on("click", "#cancel_timesheet_btn", function() {
        let name = $(this).data("name");
        frappe.call({
            method: "frappe.client.cancel",
            args: {doctype: "Timesheet", name},
            callback: function() {
                frappe.msgprint("Timesheet cancelled.");
                $("#preview_timesheet_btn").click();
            }
        });
    });

    // Create Timesheet handler
    $(document).on("click", "#create_timesheet_btn", function() {
        frappe.call({
            method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.create_timesheet_for_employee_date",
            args: get_filters(),
            callback: function(r) {
                if (r.message && r.message.success) {
                    frappe.msgprint("Timesheet created: " + r.message.timesheet);
                    $("#preview_timesheet_btn").click(); // Refresh preview
                } else {
                    frappe.msgprint(r.message && r.message.message ? r.message.message : "Failed to create timesheet.");
                }
            }
        });
    });

    // Card click handlers
    $("#card-calls").on("click", function() {
        // Fetch all projects first
        frappe.call({
            method: "frappe.client.get_list",
            args: {
                doctype: "Project",
                fields: ["name", "project_name"],
                limit_page_length: 1000
            },
            callback: function(projects_res) {
                let projects = projects_res.message || [];
                frappe.call({
                    method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.get_calls_details",
                    args: get_filters(),
                    callback: function(r) {
                        let d = r.message.details;
                        let project_summary = r.message.project_summary;
						d = d.slice().sort((a, b) => {
							if (a.project && !b.project) return -1;
							if (!a.project && b.project) return 1;
							return 0;
						});
                        let html = `
                        <div class="card p-4">
                            <h4>Calls Details</h4>
                            <div class="table-responsive mb-4">
                                <table class="table table-bordered table-hover">
                                    <thead>
                                        <tr>
                                            <th>Call DateTime</th>
                                            <th>Contact</th>
                                            <th>Link Name</th>
											<th>Client</th>
											<th>Customer No.</th>
                                            <th>Type</th>
                                            <th>Duration</th>
                                            <th style="min-width:120px">Project</th>
                                        </tr>
                                    </thead>
                                    <tbody>
									${d.map((call, idx) => {
										let projectCell = "";
										if (call.project) {
											projectCell = `<span>${call.project}</span>`;
										} else {
											let opts = `<option value=''>No Project</option>` + projects.map(p => `<option value='${p.name}' ${call.project === p.name ? 'selected' : ''}>${p.project_name || p.name}</option>`).join("");
											projectCell = `<select class=\"form-select project-select\" data-idx=\"${idx}\">${opts}</select>`;
										}
										return `
										<tr>
											<td>${call.call_datetime || ""}</td>
											<td>${call.contact || ""}</td>
											<td>${call.link_name || ""}</td>
											<td>${call.client || ""}</td>
											<td>${call.customer_no || ""}</td>
											<td>${call.calltype || ""}</td>
											<td>${call.duration || ""}</td>
											<td style=\"min-width:220px\">${projectCell}</td>
										</tr>
										`;
									}).join("")}
                                    </tbody>
                                </table>
                            </div>
                            <div class="d-flex justify-content-center mt-2">
                                <button class="btn btn-sm btn-dark" id="bulk-update-projects">Bulk Update Projects</button>
                            </div>
                            <h5 class="mt-4">Project-wise Call Time</h5>
                            <div class="table-responsive">
                                <table class="table table-bordered table-hover">
                                    <thead>
                                        <tr>
                                            <th>Project</th>
                                            <th>Total Time</th>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        ${project_summary.map(row => `
                                            <tr>
                                                <td>${row.project}</td>
                                                <td>${row.total_time}</td>
                                    </tr>
                                `).join("")}
                            </tbody>
                        </table>
                    </div>
                </div>
                `;
                $("#details-section").html(html).show();

                        // Initialize Select2 for all project dropdowns
                        setTimeout(function() {
                            $(".project-select").select2({
                                width: '100%',
                                dropdownParent: $("#details-section")
                            });
                        }, 0);

                        // Bulk update handler
                        $("#bulk-update-projects").on("click", function() {
                            let call_projects = d.map((call, idx) => {
                                return {
                                    name: call.name,
                                    project: $(".project-select[data-idx='"+idx+"']").val()
                                };
                            });
                    frappe.call({
                                method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.bulk_set_call_projects",
                                args: { call_projects: JSON.stringify(call_projects) },
                        callback: function() {
                                    frappe.msgprint("Projects updated for all calls.");
                            fetch_and_render_summary();
                                    $("#details-section").hide();
                        }
                    });
                        });
                    }
                });
            }
        });
    });

    $("#card-meetings").on("click", function() {
        frappe.call({
            method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.get_meetings_details",
            args: get_filters(),
            callback: function(r) {
                let d = r.message.details;
                let project_summary = r.message.project_summary;
                let html = `
                <div class="card p-4">
                    <h4>Meetings Details</h4>
                    <div class="table-responsive mb-4">
                        <table class="table table-bordered table-hover">
                            <thead>
                                <tr>
                                    <th>Project</th>
                                    <th>Discussion</th>
                                    <th>Arranged By</th>
                                    <th>Company Rep</th>
                                    <th>Party Rep</th>
                                    <th>Duration</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${d.map(m => `
                                    <tr>
                                        <td>${m.project || "No Project"}</td>
                                        <td>${m.discussion || ""}</td>
                                        <td>${m.arranged_by || ""}</td>
                                        <td>${m.company_representative || ""}</td>
                                        <td>${m.party_representative || ""}</td>
                                        <td>${m.duration || ""}</td>
                                    </tr>
                                `).join("")}
                            </tbody>
                        </table>
                    </div>
                    <h5>Project-wise Meeting Time</h5>
                    <div class="table-responsive">
                        <table class="table table-bordered table-hover">
                            <thead>
                                <tr>
                                    <th>Project</th>
                                    <th>Total Time</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${project_summary.map(row => `
                                    <tr>
                                        <td>${row.project}</td>
                                        <td>${row.total_time}</td>
                                    </tr>
                                `).join("")}
                            </tbody>
                        </table>
                    </div>
                </div>
                `;
                $("#details-section").html(html).show();
            }
        });
    });

    $("#card-system").on("click", function() {
        frappe.call({
            method: "productivity_next.productivity_next.page.timesheet_verificati.timesheet_verificati.get_system_details",
            args: get_filters(),
            callback: function(r) {
                let d = r.message;
                let html = `
                <div class="card p-4">
                    <h4>System Usage Project-wise</h4>
                    <div class="table-responsive">
                        <table class="table table-bordered table-hover">
                            <thead>
                                <tr>
                                    <th>Project</th>
                                    <th>Total Time</th>
                                </tr>
                            </thead>
                            <tbody>
                                ${d.map(row => `
                                    <tr>
                                        <td>${row.project || "No Project"}</td>
                                        <td>${row.total_time || "00:00"}</td>
                                    </tr>
                                `).join("")}
                            </tbody>
                        </table>
                    </div>
                </div>
                `;
                $("#details-section").html(html).show();
            }
        });
    });

    // Hide details on card click elsewhere
    $(document).on("click", function(e) {
        if (!$(e.target).closest('.summary-card, #details-section').length) {
            $("#details-section").hide();
        }
    });
};

File: productivity_next/productivity_next/page/timesheet_verificati/timesheet_verificati.py
import frappe
from frappe.utils import time_diff_in_seconds, flt
from datetime import datetime
from datetime import timedelta
from frappe.utils import today

def get_employee_and_date():
    employee = frappe.form_dict.get('employee')
    date = frappe.form_dict.get('date')
    if not date:
        date = frappe.utils.today()
    return employee, date

def secs_to_hhmm(secs):
    if not secs or secs <= 0:
        return "00:00"
    mins = int(secs // 60)
    hours = mins // 60
    mins = mins % 60
    return f"{hours:02d}:{mins:02d}"

@frappe.whitelist()
def get_summary_data():
    employee, date = get_employee_and_date()
    # Calls
    calls = frappe.db.get_all(
        'Employee Fincall',
        filters={
            'employee': employee,
            'date': date,
            'calltype': ["not in", ["Missed", "Rejected"]]
        },
        fields=['calltype', 'duration']
    )
    incoming = [c for c in calls if c['calltype'] == 'Incoming']
    outgoing = [c for c in calls if c['calltype'] == 'Outgoing']
    def sum_duration(lst):
        return sum([flt(c['duration'] or 0) for c in lst])
    incoming_dur = sum_duration(incoming)
    outgoing_dur = sum_duration(outgoing)
    total_dur = sum_duration(calls)
    # Meetings (use same SQL as details, include project)
    meetings = frappe.db.sql('''
        SELECT m.project, m.internal_meeting, m.meeting_from, m.meeting_to,
               TIMESTAMPDIFF(SECOND, m.meeting_from, m.meeting_to) as duration
        FROM `tabMeeting` m
        JOIN `tabMeeting Company Representative` mcr ON m.name = mcr.parent
        WHERE mcr.employee = %s AND m.docstatus = 1
          AND m.meeting_from >= %s AND m.meeting_to <= %s
    ''', (employee, f"{date} 00:00:00", f"{date} 23:59:59"), as_dict=True)
    internal = [m for m in meetings if m['internal_meeting']]
    external = [m for m in meetings if not m['internal_meeting']]
    def meeting_secs(lst):
        return sum([
            m['duration'] or 0
            for m in lst if m['meeting_from'] and m['meeting_to']
        ])
    internal_dur = meeting_secs(internal)
    external_dur = meeting_secs(external)
    total_meet = meeting_secs(meetings)
    # System Usage (segregate App/Web by url is null or not)
    total_web_data = frappe.db.sql(f'''
        SELECT sum(duration) as total_web_duration, count(*) as total_web_count
        FROM `tabApplication Usage log`
        WHERE date = '{date}' and employee = '{employee}' and url is not null
    ''', as_dict=True)[0]
    total_app_data = frappe.db.sql(f'''
        SELECT sum(duration) as total_app_duration, count(*) as total_app_count
        FROM `tabApplication Usage log`
        WHERE date = '{date}' and employee = '{employee}' and url is null
    ''', as_dict=True)[0]
    app_dur = flt(total_app_data["total_app_duration"] or 0)
    app_count = total_app_data["total_app_count"]
    web_dur = flt(total_web_data["total_web_duration"] or 0)
    web_count = total_web_data["total_web_count"]
    total_sys = app_dur + web_dur
    return {
        "calls": {
            "incoming": {"duration": secs_to_hhmm(incoming_dur), "count": len(incoming)},
            "outgoing": {"duration": secs_to_hhmm(outgoing_dur), "count": len(outgoing)},
            "total": secs_to_hhmm(total_dur)
        },
        "meetings": {
            "internal": {"duration": secs_to_hhmm(internal_dur), "count": len(internal)},
            "external": {"duration": secs_to_hhmm(external_dur), "count": len(external)},
            "total": secs_to_hhmm(total_meet)
        },
        "system": {
            "app": {"duration": secs_to_hhmm(app_dur), "count": app_count},
            "web": {"duration": secs_to_hhmm(web_dur), "count": web_count},
            "total": secs_to_hhmm(total_sys)
        }
    }

@frappe.whitelist()
def get_calls_details():
    employee, date = get_employee_and_date()
    calls = frappe.db.get_all(
        'Employee Fincall',
        filters={
            'employee': employee,
            'date': date,
            'calltype': ["not in", ["Missed", "Rejected"]]
        },
        fields=['name', 'client', 'duration', 'project', 'call_datetime', 'contact', 'link_name', 'calltype', 'customer_no']
    )
    details = [
        {
            'name': c['name'],
            'client': c['client'],
            'customer_no': c['customer_no'],
            'duration': secs_to_hhmm(flt(c['duration'])),
            'project': c['project'],
            'call_datetime': c['call_datetime'],
            'contact': c['contact'],
            'link_name': c['link_name'],
            'calltype': c['calltype']
        } for c in calls
    ]
    # Project-wise summary
    project_map = {}
    for c in calls:
        p = c['project'] or 'No Project'
        secs = flt(c['duration'])
        project_map.setdefault(p, 0)
        project_map[p] += secs
    project_summary = [
        {'project': p, 'total_time': secs_to_hhmm(secs)}
        for p, secs in project_map.items()
    ]
    return {
        'details': details,
        'project_summary': project_summary
    }

@frappe.whitelist()
def bulk_set_call_projects(call_projects):
    """
    call_projects: list of dicts {name: call_id, project: project_name}
    """
    import json
    call_projects = json.loads(call_projects) if isinstance(call_projects, str) else call_projects
    for row in call_projects:
        if row.get('project'):
            frappe.db.set_value("Employee Fincall", row['name'], "project", row['project'])
    frappe.db.commit()
    return True

@frappe.whitelist()
def set_call_project(call_id, project):
    frappe.db.set_value("Employee Fincall", call_id, "project", project)
    frappe.db.commit()

@frappe.whitelist()
def get_meetings_details():
    employee, date = get_employee_and_date()
    # Return project, discussion, arranged_by, company_representative, party_representative, duration
    meetings = frappe.db.sql('''
        SELECT m.project, m.discussion, m.meeting_arranged_by as arranged_by, 
               m.name as meeting_name, m.meeting_from, m.meeting_to, m.internal_meeting,
               TIMESTAMPDIFF(SECOND, m.meeting_from, m.meeting_to) as duration
        FROM `tabMeeting` m
        JOIN `tabMeeting Company Representative` mcr ON m.name = mcr.parent
        WHERE mcr.employee = %s AND m.docstatus = 1
          AND m.meeting_from >= %s AND m.meeting_to <= %s
    ''', (employee, f"{date} 00:00:00", f"{date} 23:59:59"), as_dict=True)
    for m in meetings:
        # Company reps (all employee_name from child table)
        company_reps = frappe.db.get_all(
            "Meeting Company Representative",
            filters={"parent": m["meeting_name"]},
            fields=["employee_name"]
        )
        m["company_representative"] = ", ".join([cr["employee_name"] for cr in company_reps if cr["employee_name"]])
        # Party reps (all contact from child table)
        party_reps = frappe.db.get_all(
            "Meeting Party Representative",
            filters={"parent": m["meeting_name"]},
            fields=["contact"]
        )
        m["party_representative"] = ", ".join([pr["contact"] for pr in party_reps if pr["contact"]])
        m['duration'] = secs_to_hhmm(m['duration']) if m['duration'] else "00:00"
    # Group by project for table summary
    project_map = {}
    for m in meetings:
        p = m['project'] or 'No Project'
        secs = m['duration']
        if isinstance(secs, str):
            # convert back to seconds for sum
            h, mi = [int(x) for x in secs.split(":")] if ":" in secs else (0, 0)
            secs = h * 3600 + mi * 60
        project_map.setdefault(p, 0)
        project_map[p] += secs
    project_summary = [
        {'project': p, 'total_time': secs_to_hhmm(secs)}
        for p, secs in project_map.items()
    ]
    return {
        'details': meetings,
        'project_summary': project_summary
    }
@frappe.whitelist()
def get_system_details():
    employee, date = get_employee_and_date()
    logs = frappe.db.get_all(
        'Application Usage log',
        filters={
            'date': date,
            'employee': employee
        },
        fields=['project', 'from_time', 'to_time']
    )
    project_map = {}
    for l in logs:
        if not l['project']:
            continue
        secs = time_diff_in_seconds(l['to_time'], l['from_time']) if l['from_time'] and l['to_time'] else 0
        project_map.setdefault(l['project'], 0)
        project_map[l['project']] += secs
    return [
        {'project': p, 'total_time': secs_to_hhmm(secs)}
        for p, secs in project_map.items()
    ]

@frappe.whitelist()
def preview_timesheet():
    employee, date = get_employee_and_date()
    # Find the draft or submitted timesheet created by productify for this employee and date
    timesheets = frappe.get_list(
        "Timesheet",
        filters={
            "employee": employee,
            "is_created_by_productify": 1,
            "start_date": date,
            "docstatus": ["in",[0,1]]
        },
        fields=["name", "docstatus"],
        limit=1
    )
    if not timesheets:
        return {"logs": [], "total_hours": 0, "project_summary": [], "docstatus": None, "timesheet_name": None}
    ts = frappe.get_doc("Timesheet", timesheets[0]["name"])
    logs = []
    total_hours = 0
    project_map = {}
    for row in ts.time_logs:
        logs.append({
            "activity_type": row.activity_type,
            "task": row.task,
            "project": row.project,
            "issue": row.issue,
            "hours": row.hours,
            "from_time": row.from_time,
            "to_time": row.to_time,
            "description": row.description
        })
        total_hours += row.hours or 0
        p = row.project or "No Project"
        project_map.setdefault(p, 0)
        project_map[p] += row.hours or 0
    project_summary = [
        {"project": p, "total_hours": round(h, 2)}
        for p, h in project_map.items()
    ]
    return {
        "logs": logs,
        "total_hours": round(total_hours, 2),
        "project_summary": project_summary,
        "docstatus": ts.docstatus,
        "timesheet_name": ts.name
    }

def get_employee_meetings(employee, date):
    data = frappe.db.sql(f"""
        SELECT m.name as meeting, m.meeting_from as from_time, m.meeting_to as to_time, 
           m.internal_meeting, m.project, m.task, m.meeting_arranged_by,
           mcr.employee as company_rep_employee, mcr.employee_name as company_rep_name,
            mpr.contact as party_rep_contact
        FROM `tabMeeting` as m
        LEFT JOIN `tabMeeting Company Representative` as mcr ON m.name = mcr.parent
        LEFT JOIN `tabMeeting Party Representative` as mpr ON m.name = mpr.parent
        WHERE mcr.employee = '{employee}' and m.docstatus = 1 and m.meeting_from >= '{date} 00:00:00' and m.meeting_to <= '{date} 23:59:59'
        """, as_dict=True)
    return data

@frappe.whitelist()
def create_timesheet_for_employee_date():
    employee, date = get_employee_and_date()
    # Get application logs for this employee and date
    applications = frappe.get_all(
        "Application Usage log",
        fields=["employee","from_time","to_time","task","issue","project"],
        filters={
            "employee": employee,
            "date": ["between", [date, date]],
        }
    )
    # Get call logs for this employee and date
    calls = frappe.get_all(
        "Employee Fincall",
        fields=["name as call_id","employee","call_datetime as from_time","ADDTIME(call_datetime, SEC_TO_TIME(duration)) as to_time", "issue", "task", "project"],
        filters={
            "employee": employee,
            "call_datetime": ["between", [f"{date} 00:00:00", f"{date} 23:59:59"]],
        }
    )
    # Get meeting logs for this employee and date
    meetings = get_employee_meetings(employee, date)
    # Group and merge logs for this employee
    merged_logs = {}
    applications = group_logs_by_employee(applications)
    calls = group_logs_by_employee(calls)
    merged_logs[employee] = merge_logs(applications.get(employee, []))
    merged_logs[employee] = merge_logs(merged_logs.get(employee, []) + calls.get(employee, []))
    merged_logs[employee] = split_logs(merged_logs.get(employee, []), meetings)
    merged_logs[employee] = split_logs(merged_logs.get(employee, []), calls.get(employee, []))
    # Create or update timesheet for this employee and date
    existing_timesheets = frappe.get_list(
        "Timesheet",
        filters={
            "employee": employee,
            "status": "Draft",
            "is_created_by_productify": 1,
            "start_date": date,
        },
        fields=["name"],
        limit=1
    )
    if existing_timesheets:
        timesheet = frappe.get_doc("Timesheet", existing_timesheets[0]["name"])
        timesheet.time_logs = []
    else:
        timesheet = frappe.new_doc("Timesheet")
        timesheet.employee = employee
    timesheet.is_created_by_productify = True
    timesheet.start_date = date
    employee_merged_logs = merged_logs.get(employee, [])
    if not employee_merged_logs:
        return {"success": False, "message": "No logs found for this employee and date."}
    for log in employee_merged_logs:
        log["to_time"] = log["to_time"] - timedelta(seconds=1)
        seconds = (log["to_time"] - log["from_time"]).total_seconds()
        hours = seconds / 3600
        if seconds <= 0:
            continue
        activity_type = ""
        if log.get("meeting"):
            activity_type = "Meeting"
        elif log.get("issue"):
            activity_type = "Issue"
        elif log.get("task"):
            activity_type = "Task"
        elif log.get("project"):
            activity_type = "Project"
        elif log.get("call_id"):
            activity_type = "Call"
        description = get_activity_description(log)
        timesheet.append("time_logs", {
            "activity_type": activity_type,
            "task": log.get("task"),
            "project": log.get("project"),
            "issue": log.get("issue"),
            "hours": hours,
            "from_time": log["from_time"],
            "to_time": log["to_time"],
            "description": description
        })
    try:
        if timesheet.time_logs:
            timesheet.save()
            return {"success": True, "timesheet": timesheet.name}
        else:
            return {"success": False, "message": "No logs found for this employee and date."}
    except Exception as e:
        frappe.log_error(f"Failed to create timesheet for {employee}",e)
        return {"success": False, "message": str(e)}


def group_logs_by_employee(logs):
    employee_logs = {}
    for log in logs:
        employee = log.get("employee")
        if employee not in employee_logs:
            employee_logs[employee] = []
        employee_logs[employee].append(log)
    return employee_logs


def merge_logs(logs):
    logs.sort(key=lambda x: x["from_time"])
    
    merged_logs = []
    
    for log in logs:
        if not merged_logs:
            merged_logs.append(log)
            continue
        
        last_log = merged_logs[-1]
        time_gap = (log["from_time"] - last_log["to_time"]).total_seconds()
        
        priority_keys = ["task", "issue", "project"]
        
        key_changed = any(
            log.get(key) and log[key] != last_log.get(key)
            for key in priority_keys
        )
        
        both_none = all(log.get(key) is None and last_log.get(key) is None for key in priority_keys)
        
        if time_gap <= 10 and not key_changed and (both_none or any(log.get(key) == last_log.get(key) for key in priority_keys)):
            last_log["to_time"] = max(last_log["to_time"], log["to_time"])
        
        elif log["from_time"] <= last_log["to_time"]:
            if log["to_time"] <= last_log["to_time"]:
                continue
            if log["from_time"] < last_log["to_time"]:
                last_log["to_time"] = log["from_time"]
            
            merged_logs.append(log)        
        else:
            log["from_time"] = max(log["from_time"], last_log["to_time"] + timedelta(seconds=1))
            merged_logs.append(log)
    
    return merged_logs

def split_logs(merged_logs, new_logs):
    updated_logs = []
    i, j = 0, 0

    while i < len(merged_logs) or j < len(new_logs):
        if j >= len(new_logs): 
            updated_logs.append(merged_logs[i])
            i += 1
            continue

        if i >= len(merged_logs): 
            updated_logs.append(new_logs[j])
            j += 1
            continue

        old_log = merged_logs[i]
        new_log = new_logs[j]

        if new_log["to_time"] <= old_log["from_time"]:
            updated_logs.append(new_log)
            j += 1
        elif new_log["from_time"] >= old_log["to_time"]:
            updated_logs.append(old_log)
            i += 1
        else:
            if old_log["from_time"] < new_log["from_time"]:
                updated_logs.append({
                    **old_log,
                    "to_time": new_log["from_time"]
                })
            updated_logs.append(new_log)
            j += 1
            
            if old_log["to_time"] > new_log["to_time"]:
                merged_logs[i]["from_time"] = new_log["to_time"]
            else:
                i += 1
    for k in range(1, len(updated_logs)):
        if updated_logs[k]["from_time"] < updated_logs[k - 1]["to_time"]:
            updated_logs[k]["from_time"] = updated_logs[k - 1]["to_time"]
    return updated_logs

def get_activity_description(activity):
    """Generate description using Task and Issue subject lines."""
    description_parts = []

    if activity.get('task'):
        task_subject = frappe.db.get_value("Task", activity['task'], "subject") or ""
        if task_subject:
            description_parts.append(f"Task - {task_subject}")

    if activity.get('issue'):
        issue_subject = frappe.db.get_value("Issue", activity['issue'], "subject") or ""
        if issue_subject:
            description_parts.append(f"Issue - {issue_subject}")

    return "\n".join(description_parts)

@frappe.whitelist()
def submit_timesheet(name):
    doc = frappe.get_doc("Timesheet", name)
    doc.submit()
    return {"success": True}